Add guidelines for making F# data types structs (#16699)
dotnet/docs: docs/fsharp/style-guide/conventions.md

## Performance

### Prefer structs for small data types

Using structs (also called Value Types) can often result in higher performance for some code because it typically avoids allocating objects. However, structs are not always a "go faster" button: if the size of the data in a struct exceeds 16 bytes, copying the data can often result in more CPU time spend than using a reference type.

> To determine if you should use a struct, consider the following two points:

Reviewer: not sure if you need to number them

```suggestion
To determine if you should use a struct, consider the following conditions:
```

> 1. If the size of your data is 16 bytes or smaller

Reviewer: ```suggestion
- If the size of your data is 16 bytes or smaller.
```

> 2. If you are likely to have many of these data types resident in memory in a running program

Reviewer: ```suggestion
- If you're likely to have many of these data types resident in memory in a running program.
```

> If the first point applies, you should generally use a struct. If both apply, you should almost always use a struct.

Reviewer: ```suggestion
If the first condition applies, you should generally use a struct. If both apply, you should almost always use a struct.
```

> 

Reviewer: Should you explain why almost always and not always?

Author: Yep, did. Gist of it is that if you measure, you could still find no difference or perhaps structs being worse. So it's always important to measure.

#### Prefer struct tuples when grouping small value types

Consider the following two functions:

```fsharp
let rec runWithTuple t offset times =
    let offsetValues x y z offset =
        (x + offset, y + offset, z + offset)

    if times <= 0 then
        t
    else
        let (x, y, z) = t
        let r = offsetValues x y z offset
        runWithTuple r offset (times - 1)

let rec runWithStructTuple t offset times =
    let offsetValues x y z offset =
        struct(x + offset, y + offset, z + offset)

    if times <= 0 then
        t
    else
        let struct(x, y, z) = t
        let r = offsetValues x y z offset
        runWithStructTuple r offset (times - 1)
```

When you benchmark these functions with a statistical benchmarking tool like [BenchmarkDotNet](https://benchmarkdotnet.org/), you'll find that the `runWithStructTuple` function that uses struct tuples runs 40% faster and allocates no memory.

> However, these results will not always be the case in your own code. Depeneding on if you mark a function as `inline`, code that uses reference tuples may get some additional optimizations, or code that would allocate could simply be optimized away. You should always measure results whenever performance is concerned, and never operate based on assumption or intuition.

Reviewer: ```suggestion
However, these results won't always be the case in your own code. If you mark a function as `inline`, code that uses reference tuples may get some additional optimizations, or code that would allocate could simply be optimized away. You should always measure results whenever performance is concerned, and never operate based on assumption or intuition.
```

#### Prefer struct records when the data type is small

The rule of thumb described earlier also holds for [F# record types](../language-reference/records.md). Consider the following data types and functions that process them:

```fsharp
type Point = { X: float; Y: float; Z: float }

[<Struct>]
type SPoint = { X: float; Y: float; Z: float }

let rec processPoint (p: Point) offset times =
    let inline offsetValues (p: Point) offset =
        { p with X = p.X + offset; Y = p.Y + offset; Z = p.Z + offset }

    if times <= 0 then
        p
    else
        let r = offsetValues p offset
        processPoint r offset (times - 1)

let rec processStructPoint (p: SPoint) offset times =
    let inline offsetValues (p: SPoint) offset =
        { p with X = p.X + offset; Y = p.Y + offset; Z = p.Z + offset }

    if times <= 0 then
        p
    else
        let r = offsetValues p offset
        processStructPoint r offset (times - 1)
```

> This is very similar to the tuple code before, but this time using records and an inlined inner function.

Reviewer: ```suggestion
This is very similar to the previous tuple code, but this time the example uses records and an inlined inner function.
```

When you benchmark these functions with a statistical benchmarking tool like [BenchmarkDotNet](https://benchmarkdotnet.org/), you'll find that `processStructPoint` runs nearly 60% faster and allocates nothing on the managed heap.

#### Prefer struct discriminated unions when the data type is small

The previous observations about performance with struct tuples and records also holds for [F# Discriminated Unions](../language-reference/discriminated-unions.md). Consider the following code:

```fsharp
    type Name = Name of string
    
    [<Struct>]
    type SName = SName of string

    let reverseName (Name s) =
        s.ToCharArray()
        |> Array.rev
        |> string
        |> Name

    let structReverseName (SName s) =
        s.ToCharArray()
        |> Array.rev
        |> string
        |> SName
```

> It's very common to define single-case Discriminated Unions like this for domain modeling. When you benchmark these functions with a statistical benchmarking tool like [BenchmarkDotNet](https://benchmarkdotnet.org/) you'll find that `structReverseName` runs about 25% faster than `reverseName` for small strings. For very large strings, both will perform about the same, so in this case it is always preferable to use a struct. As mentioned before, always measure and do not operate on assumptions or intuition.

Reviewer: ```suggestion
It's very common to define single-case Discriminated Unions like this for domain modeling. When you benchmark these functions with a statistical benchmarking tool like [BenchmarkDotNet](https://benchmarkdotnet.org/), you'll find that `structReverseName` runs about 25% faster than `reverseName` for small strings. For very large strings, both perform about the same. So, in this case, it's always preferable to use a struct. As previously mentioned, always measure and do not operate on assumptions or intuition.
```

Although the previous example showed that a struct Discriminated Union yielded better performance, it is quite common to have larger Discriminated Unions when modeling a domain. Larger data types like that may not perform as well if they are structs depending on the operations on them, since more copying could be involved.

### Functional programming and mutation

F# values are immutable by default, which allows you to avoid certain classes of bugs (especially those involving concurrency and parallelism). However, in certain cases, in order to achieve optimal (or even reasonable) efficiency of execution time or memory allocations, a span of work may best be implemented by using in-place mutation of state. This is possible in an opt-in basis with F# with the `mutable` keyword.

> Use of `mutable` in F# may feel at odds with functional purity. This is understandable, but functional purity everywhere can be at odds with performance goals. A compromise is to encapsulate mutation such that callers need not care about what happens when they call a funciton. This allows you to write a functional interface over a mutation-based implementation for performance critical code.

Reviewer: F# has some funny names so not sure if funciton is one of them :-) 

```suggestion
Use of `mutable` in F# may feel at odds with functional purity. This is understandable, but functional purity everywhere can be at odds with performance goals. A compromise is to encapsulate mutation such that callers need not care about what happens when they call a function. This allows you to write a functional interface over a mutation-based implementation for performance-critical code.
```

## Performance

### Prefer structs for small data types

Using structs (also called Value Types) can often result in higher performance for some code because it avoids allocating and de-allocating objects on the managed heap in .NET. However, structs are not always a "go faster" button: if the size of the data in a struct exceeds 16 bytes, copying the data can often result in more CPU time spend than using a reference type.

A good rule of thumb for using structs intead of value types is:

* If the size of your data is 16 bytes or smaller (measure with a call to `sizeof`)

> * If your data type is made up of other value types (e.g., a record of `int`s)

Reviewer 2: ```suggestion
* If your data type is made up of other value types (for example, a record of `int`s)
```

https://docs.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/e/eg
